sprint_6-solution

RomaB1983 opened the pull request.

Commit:
Спринт 6. Техническое задание
     - связанный список(на основе LinkedList)
     - тесты

Ivanov-s-tmn commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,29 +3,84 @@
+        }
+    }
+
+    /**
+     * В ТЗ написано, что надо из двусвязногосписка формировать список таск.
+     * Можешь пояснить зачем? Почему нельзя собрать их из history через values(), и также в цикле собрать List?
+     * На мой взгляд выглядит логичнее. Так мы точно соберем то, что в истории..
+     * Вроде и не дольше должна работать такая реализация..
+    */
```
`values()` не гарантирует порядок элементов, в хешмапе местоположение элемента в памяти определяется по хешкоду ключа и оно может быть произвольным

Ivanov-s-tmn commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,29 +3,84 @@
+    @Override
+    public void remove(int id) {
+        removeNode(history.get(id));
+        history.remove(id);
```
Перенеси удаление ноды из хешмапы в метод `removeNode() `, тут нода удаляется из хешмапы, а в методе `add()` нет

RomaB1983 replied:
Мне показалось, что в методе add это было бы лишним. Поскольку я следующим шагом перезаписываю значение ноды: history.put(task.getId(), linkLast(task)). Получается. что history.remove(id) тут особо и не нужен. Или я ошибаюсь?

Ivanov-s-tmn replied:
Да, ты прав

Ivanov-s-tmn commented on src/service/InMemoryTaskManager.java:
В методах удаления всех подзадач/задач/эпиков тоже удаляй из истории

RomaB1983 replied:
Поправил. Пришлось добавить метод remove(Collection<Integer>)

Ivanov-s-tmn commented on src/service/Node.java:
Класс можно сделать вложенным в `InMemoryTaskManager` (private static), он нужен только там и нет особого смысла выносить ег ов отдельный класс. Так же сделано и в реализации `LinkedList`

RomaB1983 replied:
Поправил

Ivanov-s-tmn commented on test/service/InMemoryHistoryManagerTest.java:
Обрати внимание, что не проходят 2 теста, поправь их)

Еще, так как мы реализовали историю двусвязным списком, нужно покрыть тестами следующие случаи (удаление из начала/из середины/из конца истории)

RomaB1983 replied:
поправил

Ivanov-s-tmn commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,29 +3,84 @@
+    }
+
+    private Node linkLast(Task task) {
+        final Node last = tail;
```
Можно не записывать в отдельную переменную `last` предыдущую последнюю ноду `tail`. 
Создавай сразу новую ноду с ссылкой на `tail`, а `tail = newNode;` вынеси в конец метода, плюс в else блоке можно тоже работать с `tail.next = newNode;`

RomaB1983 replied:
Поправил

Ivanov-s-tmn commented on src/service/InMemoryHistoryManager.java:
Отлично реализован двусвязный список и методы добавления/удаления/обхода нод!

Commit:
Спринт 6. Этап1. Устранение замечаний:
- перенес класс Node в InMmoryHistoryManager
- поправил метод LinkLast
- добавил метод remove(Collection) в HistoryManager для удаления по спиcку id
- попрвил тесты

Commit:
Спринт 6. Этап1. Устранение замечаний:
- исправление checkStyle

Ivanov-s-tmn commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -26,8 +26,17 @@ public void add(Task task) {
+    public void remove(Collection<Integer> ids) {
+        for (int id : ids) {
+            remove(id);
+        }
```
У менеджера истории должна быть возможность удаления только одной задачи из истории, удаляй в циклах или в стримах в методах таск менеджера

RomaB1983 replied:
поправил

Ivanov-s-tmn commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -46,4 +46,43 @@ void isShoulbeOneTaskInHistoryWhenAddTaskTwice() {
 
+    @Test
+    void isShoulbeRemoveInHistoryWhenRemoveLast() {
+        Task task = new Task("таск ", "таск " );
+        Task task2 = new Task("таск2 ", "таск2 " );
+        taskManager.addTask(task);
+        taskManager.addTask(task2);
+        taskManager.getTask(task2.getId());
+        taskManager.getTask(task.getId());
+        taskManager.deleteTask(task2.getId());
+        assertEquals(task, taskManager.getHistory().get(0));
+    }
+
+    @Test
+    void isShoulbeRemoveInHistoryWhenRemoveFirst() {
+        Task task = new Task("таск ", "таск " );
+        Task task2 = new Task("таск2 ", "таск2 " );
+        taskManager.addTask(task);
+        taskManager.addTask(task2);
+        taskManager.getTask(task2.getId());
+        taskManager.getTask(task.getId());
+        taskManager.deleteTask(task.getId());
+        assertEquals(task2, taskManager.getHistory().get(0));
+    }
+
+    @Test
+    void isShoulbeRemoveInHistoryWhenRemoveMedium() {
+        Task task = new Task("таск ", "таск " );
+        Task task2 = new Task("таск2 ", "таск2 " );
+        Task task3 = new Task("таск3 ", "таск3 " );
+        taskManager.addTask(task);
+        taskManager.addTask(task2);
+        taskManager.addTask(task3);
+        taskManager.getTask(task.getId());
+        taskManager.getTask(task2.getId());
+        taskManager.getTask(task3.getId());
+        taskManager.deleteTask(task2.getId());
+        assertEquals(task, taskManager.getHistory().get(0));
+        assertEquals(task3, taskManager.getHistory().get(1));
+    }
```
Эти тесты лучше проверять с одинаковыми начальными условиями, как в 3 тесте - 3 задачи в истории. Вынеси создание и добавление задач в историю в метод beforeEach, чтобы это не дублировать в тестовых методах. 
Так как это тесты для менеджера истории, работай только с его методами, таск менеджер ту тне нужен.

RomaB1983 replied:
Поправил

Commit:
Спринт 6. Этап2. Устранение замечаний:
- убрал taskManager из тестов для historyManager
- убрал remove(Collection<Integer>) из historyManager

Ivanov-s-tmn commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -4,85 +4,66 @@
-        taskManager.deleteTask(task2.getId());
-        assertEquals(task, taskManager.getHistory().get(0));
+        historyManager.remove(task2.getId());
+        assertEquals(task, historyManager.getHistory().get(0));
```
Последняя задач - это 3. При удалении 3 задачи в списке должна остаться 1 и 2, добавь ассерт

RomaB1983 replied:
Поправил (слажал маленько))

Ivanov-s-tmn commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -4,85 +4,66 @@
-        taskManager.getTask(task.getId());
-        taskManager.deleteTask(task.getId());
-        assertEquals(task2, taskManager.getHistory().get(0));
+        historyManager.remove(task.getId());
+        assertEquals(task2, historyManager.getHistory().get(0));
```
тут тоже в списке должна остаться 2 и 3

RomaB1983 replied:
Поправил (слажал маленько))

Commit:
Спринт 6. Этап3. Устранение замечаний:
- поправил некорректные тесты в HistoryManagerTest
